kyo-prelude/shared/src/main/scala/kyo2/kernel/Safepoint.scala

```diff
@@ -0,0 +1,121 @@
+
+    private[kernel] def clearTrace(): Unit =
+        traceIdx = 0
+
+    private def exit(depth: Int): Unit =
+        this.depth = depth - 1
+
+    private[kernel] def save(context: Context) =
+        State(Safepoint.copyTrace(trace, traceIdx), traceIdx, context)
+end Safepoint
+
+object Safepoint:
+
+    implicit def get: Safepoint = local.get()
+
+    private[kernel] inline def eval[T](
+        inline f: => T
+    )(using inline frame: Frame): T =
+        val parent = Safepoint.local.get()
+        val self   = new Safepoint(0, State.empty)
+        Safepoint.local.set(self)
+        self.pushFrame(frame)
+        try f
+        catch
+            case ex: Throwable if NonFatal(ex) =>
+                handle(self, ex)
+        finally
+            Safepoint.local.set(parent)
+        end try
+    end eval
+
+    private[kernel] inline def handle[A, S](
+        inline suspend: Safepoint ?=> A < S,
+        inline continue: => A < S
+    )(using inline frame: Frame, self: Safepoint): A < S =
+        self.enter(frame) match
+            case -1 => Effect.defer(suspend)
+            case depth =>
+                try continue
+                finally self.exit(depth)
+    end handle
+
+    private def handle(self: Safepoint, cause: Throwable): Nothing =
+        val size  = Math.min(self.traceIdx, maxTraceFrames)
+        val trace = copyTrace(self.trace, self.traceIdx).filter(_ != null).map(_.parse)
+        val toPad = trace.map(_.snippetShort.size).max + 1
+        val elements = trace.map { frame =>
+            StackTraceElement(
+                frame.snippetShort.reverse.padTo(toPad, ' ').reverse + " @ " + frame.declaringClass,
+                frame.methodName,
+                frame.position.fileName,
+                frame.position.lineNumber
+            )
+        }.reverse
+        val prefix = cause.getStackTrace.takeWhile(e =>
+            e.getFileName() != elements(0).getFileName() || e.getLineNumber != elements(0).getLineNumber()
+        )
+        val suffix = (new Exception).getStackTrace().drop(2)
+        cause.setStackTrace(prefix ++ elements ++ suffix)
+        throw cause
+    end handle
+
+    private def copyTrace(trace: Array[Frame], idx: Int): Array[Frame] =
+        val result = new Array[Frame](maxTraceFrames)
+        val offset = Math.max(0, idx - maxTraceFrames)
+        @tailrec def loop(i: Int): Unit =
+            if i < maxTraceFrames then
+                val t = trace((offset + i) & (maxTraceFrames - 1))
+                if t != null then
+                    result(i) = t
+                    loop(i + 1)
+        loop(0)
+        result
+    end copyTrace
+
+    import internal.*
+
+    final class State(
+        val trace: Array[Frame],
+        val traceIdx: Int,
+        val context: Context
+    )
+
+    object State:
+        val empty = State(new Array(maxStackDepth), 0, Context.empty)
+
+    private[kernel] val local = ThreadLocal.withInitial(() => Safepoint(0, State.empty))
```

Comment:
Do you expect any serious performance limitation of Thread locals? I wonder if `ScopedValue` could be used here - though that limits the executor to newer JDKs.

Reply:
Interesting, I thought `ScopedValue` was just a scoped wrapper for `ThreadLocal` but it's actually a more efficient mechanism. I'll play with it, if the performance is better we might be able to avoid the optimization to propagate `Safepoint` via implicits.

Reply:
I took a look at `ScopedValue` and it doesn't seem possible to use it to store the `Safepoint` since its use isn't always scoped. For example, if a computation is executed without any suspensions, it'll get the safepoint from the thread local even if the code isn't within an `.eval` execution. We should review this later, though. We might need a more involved solution when the scheduler's Loom integration is enabled since the thread local increases the memory footprint of virtual threads, and there can be many of them.